java-components/cache/src/main/java/com/redhat/hacbs/artifactcache/services/client/ociregistry/OCIRegistryRepositoryClient.java

```diff
@@ -0,0 +1,229 @@
+        } finally {
+            Log.warnf("OCI registry request to %s:%s:%s took %sms", group, artifact, version,
+                    System.currentTimeMillis() - time);
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<RepositoryResult> getMetadataFile(String buildPolicy, String group, String target) {
+        return Optional.empty();
+    }
+
+    private RegistryClient getRegistryClient(String registry, String owner, String group, String hashedGav,
+            boolean enableHttpAndInsecureFailover) {
+        ImageReference imageReference = ImageReference.of(registry, owner + File.separator + group, hashedGav);
+        CredentialRetrieverFactory credentialRetrieverFactory = CredentialRetrieverFactory.forImage(imageReference,
+                (s) -> Log.info(s.getMessage()));
+
+        RegistryClient.Factory factory = RegistryClient.factory(new EventHandlers.Builder().build(), registry,
+                owner + File.separator + group,
+                new FailoverHttpClient(enableHttpAndInsecureFailover, false, s -> Log.info(s.getMessage())));
+
+        try {
+            Optional<Credential> credential = credentialRetrieverFactory.dockerConfig().retrieve();
+            if (credential.isPresent()) {
+                factory.setCredential(credential.get());
+            } else {
+                Log.warn("No credential found for " + registry + ", proceeding without any");
+            }
+        } catch (CredentialRetrievalException cre) {
+            throw new RuntimeException(cre);
+        }
+
+        RegistryClient registryClient = factory.newRegistryClient();
+
+        return registryClient;
+    }
+
+    private Path getLocalCachePath(RegistryClient registryClient, ManifestTemplate manifest, String digestHash)
+            throws IOException {
+        if (imageCache.containsKey(digestHash)) {
+            return pullStale(registryClient, manifest, digestHash);
+        } else {
+            return pullFreshAndCache(registryClient, manifest, digestHash);
+        }
+    }
+
+    private Path pullStale(RegistryClient registryClient, ManifestTemplate manifest, String digestHash) throws IOException {
+        Path path = imageCache.get(digestHash);
+        if (Files.exists(path)) {
+            return path;
+        } else {
+            return pullFreshAndCache(registryClient, manifest, digestHash);
+        }
+
+    }
+
+    private Path pullFreshAndCache(RegistryClient registryClient, ManifestTemplate manifest, String digestHash)
+            throws IOException {
+
+        String manifestMediaType = manifest.getManifestMediaType();
+
+        if (OCI_MEDIA_TYPE.equalsIgnoreCase(manifestMediaType)) {
+            List<BuildableManifestTemplate.ContentDescriptorTemplate> layers = ((OciManifestTemplate) manifest).getLayers();
+
+            // Layer 2 is artifacts
+            BuildableManifestTemplate.ContentDescriptorTemplate artifactsLayer = layers.get(2);
```

Comment:
We probably need some error handling here.

Reply:
Meaning ? If there is not a layer 2 ?

Reply:
yea, if the image is not valid it should probably log an error and return.

Reply:
Cool. Adding that.